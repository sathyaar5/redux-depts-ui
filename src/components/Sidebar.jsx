import React, { useState } from "react";
import {
  List,
  ListItem,
  ListItemText,
  Typography,
  Button,
  TextField,
  Grid,
} from "@mui/material";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";

import { setSelectedDepartment, addDepartment } from "../actions";

const Sidebar = () => {
  const departments = useSelector((state) => state.departments);
  const dispatch = useDispatch();
  const [showAddDepartmentForm, setShowAddDepartmentForm] = useState(false);
  const [departmentName, setDepartmentName] = useState("");
  const [managerName, setManagerName] = useState("");
  const [errors, setErrors] = useState({});

  const handleDepartmentClick = (index) => {
    dispatch(setSelectedDepartment(index));
  };

  const handleAddDepartment = () => {
    setShowAddDepartmentForm(true);
  };

  const handleSubmit = () => {
    if (!departmentName.trim()) {
      setErrors({ departmentName: "Department name is required" });
      return;
    }
    if (!managerName.trim()) {
      setErrors({ managerName: "Manager name is required" });
      return;
    }

    setErrors({});
    dispatch(addDepartment(departmentName, managerName));

    setShowAddDepartmentForm(false);
    setDepartmentName("");
    setManagerName("");
  };

  return (
    <div>
      <Typography variant="h5" gutterBottom>
        Departments
      </Typography>
      <List>
        {departments &&
          departments.map((department, index) => (
            <ListItem
              key={index}
              button
              onClick={() => handleDepartmentClick(index)}
              className="department-item"
            >
              <Link
                to={`/${department.deptName
                  .toLowerCase()
                  .replace(/\s+/g, "-")}`}
              >
                <ListItemText primary={department.deptName} />
              </Link>
            </ListItem>
          ))}
      </List>
      {showAddDepartmentForm ? (
        <Grid container spacing={2} alignItems="center">
          <Grid item xs={12}>
            <TextField
              fullWidth
              placeholder="Department Name"
              data-testid="department-name-input"
              value={departmentName}
              onChange={(e) => setDepartmentName(e.target.value)}
              error={!!errors.departmentName}
              helperText={errors.departmentName}
            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              fullWidth
              placeholder="Manager Name"
              data-testid="manager-name-input"
              value={managerName}
              onChange={(e) => setManagerName(e.target.value)}
              error={!!errors.managerName}
              helperText={errors.managerName}
            />
          </Grid>
          <Grid item xs={12}>
            <Button variant="contained" onClick={handleSubmit}>
              Add
            </Button>
          </Grid>
        </Grid>
      ) : (
        <Button variant="contained" onClick={handleAddDepartment}>
          Add Department
        </Button>
      )}
    </div>
  );
};

export default Sidebar;
