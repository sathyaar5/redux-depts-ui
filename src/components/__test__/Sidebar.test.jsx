import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import store from "../../store";
import Sidebar from "../Sidebar";

test("renders department names correctly", () => {
  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    </Provider>,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("allows adding new department", async () => {
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Sidebar />
      </MemoryRouter>
    </Provider>,
  );

  const addDepartmentButton = screen.getByText("Add Department");
  fireEvent.click(addDepartmentButton);

  const departmentNameInput = screen.getByLabelText("Department Name");
  const managerNameInput = screen.getByLabelText("Manager Name");

  fireEvent.change(departmentNameInput, {
    target: { value: "Marketing Department" },
  });
  fireEvent.change(managerNameInput, { target: { value: "Sathya" } });

  const addButton = screen.getByText("Add");
  fireEvent.click(addButton);

  // Wait for any asynchronous operations to complete
  await screen.findByText("Marketing Department");

  const tree = renderer
    .create(
      <Provider store={store}>
        <MemoryRouter>
          <Sidebar />
        </MemoryRouter>
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});