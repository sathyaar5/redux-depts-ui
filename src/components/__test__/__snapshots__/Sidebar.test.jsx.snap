// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows adding new department 1`] = `
<div>
  <h5
    className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
  >
    Departments
  </h5>
  <ul
    className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <div
      className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button department-item css-bshv44-MuiButtonBase-root-MuiListItem-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <a
        href="/marketing-department"
        onClick={[Function]}
      >
        <div
          className="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Marketing Department
          </span>
        </div>
      </a>
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </ul>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Add Department
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`renders department names correctly 1`] = `
<div>
  <h5
    className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
  >
    Departments
  </h5>
  <ul
    className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  />
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Add Department
  </button>
</div>
`;
